pkg/k8s/watchers/watcher.go

```diff
@@ -586,6 +568,59 @@ func (k *K8sWatcher) addK8sSVCs(svcID k8s.ServiceID, svc *k8s.Service, endpoints
 			svcs = append(svcs, dpSVC...)
 		}
 	}
+	return svcs
+}
+
+// hashSVCMap returns a mapping of all frontend's hash to the its corresponded
+// value.
+func hashSVCMap(svcs []loadbalancer.SVC) map[string]loadbalancer.L3n4Addr {
+	m := map[string]loadbalancer.L3n4Addr{}
+	for _, svc := range svcs {
+		m[svc.Frontend.L3n4Addr.Hash()] = svc.Frontend.L3n4Addr
+	}
+	return m
+}
+
+func (k *K8sWatcher) addK8sSVCs(svcID k8s.ServiceID, oldSvc, svc *k8s.Service, endpoints *k8s.Endpoints) error {
+	// If east-west load balancing is disabled, we should not sync(add or delete)
+	// K8s service to a cilium service.
+	if option.Config.DisableK8sServices {
+		return nil
+	}
+
+	// Headless services do not need any datapath implementation
+	if svc.IsHeadless {
+		return nil
+	}
+
+	scopedLog := log.WithFields(logrus.Fields{
+		logfields.K8sSvcName:   svcID.Name,
+		logfields.K8sNamespace: svcID.Namespace,
+	})
+
+	svcs := datapathSVCs(svc, endpoints)
+	svcMap := hashSVCMap(svcs)
+
+	if oldSvc != nil {
+		// If we have oldService then we need to detect which frontends
+		// are no longer in the updated service and delete them in the datapath.
+
+		oldSVCs := datapathSVCs(oldSvc, endpoints)
+		oldSVCMap := hashSVCMap(oldSVCs)
+
+		for svcHash, oldSvc := range oldSVCMap {
```

Comment:
How does this behave in terms of nodeport services where we have 3 frontend entries per service which k8s may not be aware about, would we remove e.g. the wildcard entry before re-adding them and thus create a race in the datapath?

Reply:
Discussed offline, all good to proceed.